### ISSUE TITLE
DecisionTree: minimumLeafSize dramatically affects scores

### ISSUE THREAD

**Issue by rayrapetyan** [mlpack/mlpack#1399]
Looks like changing minimumLeafSize when building a decision tree dramatically affects accuracy, recall etc. E.g. here are few examples:

min_leaf_size,accuracy,recall
30,0.97,0.706
230,0.97,0.69
280,0.95,0.99

Or another dataset:
min_leaf_size,accuracy,recall
100,0.97,0.82
110,0.94,0.99

See - the difference of just 10 gives a totally different recall. It's hard to predict which value will give us a best recall, or there are some approaches for that?

From the other side, Python's scikit-learn gives the best recall from the first try and changing max_depth and other input params doesn't produce such huge differences on the same data set.

Thanks.

**Comment by rcurtin** [comment]
Hi there @rayrapetyan, thanks for getting in touch.  I can't speak for exactly what scikit is doing, but this implementation is probably most similar to Weka's J48 implementation.  (It's basically the ID3 algorithm.)  Unfortunately based on the information you've given, I can't really diagnose whether or not something is wrong.  My immediate thoughts seeing your results and report are:

 * scikit-learn's `max_depth` parameter is pretty different than `minimumLeafSize` in mlpack.  I'd say that it's closer to the parameters `min_samples_split` and `min_samples_leaf`.  It looks like with default options, it will give something close to what mlpack will give with `minimumLeafSize = 1`.

 * How are you calculating accuracy and recall?  Note that these both depend on the thresholds you are choosing, and the right threshold may be different for different trees.  Really the better measure here would be either the AUC or to look at an ROC curve.

 * Is the accuracy and recall being calculated on a holdout test set or on the training set?  A decision tree with `minimumLeafSize = 1` should be able to perfectly fit the training set (100% accuracy), but is likely to overfit and thus perform poorly on the test set.

I would say that mlpack's decision trees are pretty well tested, so I don't expect that there is a bug here, but I'd need some more information to be able to figure out what is going on here (i.e. whether or not this is expected behavior, or if something is wrong).

**Comment by djently** [comment]
Hi @rcurtin, it's not clear what do you mean by the threshold for decision trees. As far as I know, the decision rule for a terminal node in decision trees is simple majority. I can suppose that we can use a threshold for proportion of target class samples in leaf, but I've never seen such implementation in any machine learning libraries.

BTW,  CART algorithm is implemented for decision trees in scikit-learn. I'm not sure, but maybe ID3 is just more sensitive to minimumLeafSize parameter.

**Comment by rcurtin** [comment]
scikit's implementation provides `predict_proba()` and mlpack provides a `Classify()` overload that returns class probabilities.  In a typical setting, you would take these class probabilities and then create an ROC curve in order to select a threshold that best reflects the performance tradeoff you are looking for (between TPR/FPR, or precision/recall, either way).

ID3 was actually only originally designed for categorical features, whereas CART can support both categorical and numeric features.  There are a lot of connections between the two techniques, and I can't think of any extra sensitivity to the minimum leaf size parameter, if scikit's implementation is just regular CART.

If you plot an ROC curve, that will tell us much more about what is going on here and whether there actually is a problem, or if the threshold for classification just needs to be tuned.  At this point I am not certain that there is an issue, but once we see an ROC curve we will know more.

**Comment by rayrapetyan** [comment]
@rcurtin, please find ROC curves below. Thanks.
![roc](https://user-images.githubusercontent.com/4700523/40569495-25272b7c-6026-11e8-94b9-f7fd1b35fbb9.png)

**Comment by rcurtin** [comment]
@rayrapetyan: thanks for taking the time to create those, this doesn't look the way I would expect (I'd have expected these to be roughly the same).  Let me do a little experimentation to see if the same trend holds for other datasets, and I'll update this issue with what I find.

**Comment by rcurtin** [comment]
Hi @rayrapetyan, I'm sorry this response took so long.  The past weeks have been very busy and I was not able to devote the time.

I am not able to reproduce the issue that you've shown.  I took your data and took a look.  The training data is roughly 11k points and the test data is much larger, about 4M points.  This test data shows huge label skew towards label 0, and I think this may be a part of the strange effects you might be seeing.  Sometimes very unbalanced datasets can produce strange results.  However, I was not able to reproduce any strange results.  I trained models under several conditions:

 * I trained four mlpack models with minimum leaf sizes of 1, 5, 10, and 20.
 * I trained scikit models: one with default parameters, one with min_samples_split of 10, and one with max_depth of 3.

I then plotted these on ROC curves.  Note that the curve is different depending on whether class 0 or class 1 is the positive class, so I plotted both.  They look roughly the same.

http://www.ratml.org/misc_img/roc_curves_0.png
http://www.ratml.org/misc_img/roc_curves_1.png

I want to make a few points about these results:

  1. It does not appear that mlpack is significantly underperforming scikit, and it appears that it is relatively easy to tune minimum_leaf_size: if it is too small, the tree overfits (orange line); if it is too large, the tree underfits (black line).
  2. Even scikit with default parameters does not do well, since the default is minimum_samples_split = 1.  It is tough to see but the red line overlaps the orange line basically exactly.  When I tried with other datasets, with default parameters I also found that mlpack and scikit gave basically identical results when I ran mlpack with minimum_leaf_size = 1 and scikit with default parameters.
  3. The tuning of the threshold makes a huge difference for the performance of the classifier.  When you do majority classification like you suggested, this produces the "X"s that I have plotted.  Note that that X is only _one_ point on the ROC curve that you could operate at.  So it is possible to choose a different threshold to achieve a different TPR/FPR rate.  You were looking at accuracy, TNR, and precision, but it is possible to tune those also using a precision/recall curve in much the same way (I simply prefer ROC curves because I am more used to them, no particular reason).
  4. I can't reproduce anything quite like the ROC curve that you plotted above.  I am not sure exactly what is different between the two but I am highly confident that the script I ran to generate my figures is correct given your data.

Overall, after running all these simulations, my conclusion is that there is no bug in mlpack's implementation, and that minimum_leaf_size is not a difficult-to-tune parameter.  Your original observation, that minimum_leaf_size dramatically affects scores, is correct and expected, but like I said the tradeoff between minimum_leaf_size, overfitting, and underfitting is pretty straightforward.  I see the same behavior in scikit, and I'm sure I would if I used different toolkits also.

The script that I wrote to generate the figures is found at http://www.ratml.org/misc/plot_rocs.py.  It uses mlpack in Python and assumes your data is in the `data/` subdirectory of the current working directory.

**Comment by rayrapetyan** [comment]
@rcurtin, FYI the training data is a sampled data, that is all positive and same amount of negative predictions, that's why it's skewed, which is expected...

**Comment by rcurtin** [comment]
That's fine, as long as you know the data is skewed that's good. I just wanted to make sure that you knew, since it was an observation I had.  Nonetheless, my conclusion is still that I can't find any mlpack bug or reason to implement/prefer a max_depth tuning parameter.  Since I see no issues in the simulations I am running, I'd urge you to check that there aren't any issues with your code or how you are calculating your results or anything, and I'd also suggest an ROC-curve-based approach for evaluating your classifier (or something like this, precision-recall curve is fine too like I said).

**Comment by rcurtin** [comment]
Hi @rayrapetyan---I don't mean to be rude or anything but I see in the daal issue that referenced this one you claimed mlpack produced weak predictions.  I spent quite some time running simulations to confirm if there was any performance issue and concluded (as you can see in my writeup above) that there is not, and that I can achieve virtually the exact same performance as scikit.  All you need to do is tune the threshold for classification for your desired TPR/FPR rate and I think you will find the performance you are seeking.

I am happy to help you out and explain further if anything I wrote is unclear; I'm convinced that mlpack is able to provide the performance that you are looking for (and I can see that you are putting in a lot of effort trying to find some solution for FreeBSD).  At the same time, I don't want to try and coerce you into using mlpack if you don't want to.  It just seems like it would solve your problem just fine based on what I have seen. :)

**Comment by rayrapetyan** [comment]
@rcurtin, you're right, sorry I haven't performed any research after your last response, mlpack indeed gives comparable result after tuning (and that's what I've mentioned in the very first message in this topic btw). But from what I've experienced in practice (and provided diagrams above) with different data sets - skiit-learn always outperforms mlpack on our data sets with a default tree width. Maybe it's specific for our data sets, maybe mlpack compiled with different versions of fortran (gfortran vs flang used in my custom built armadillo) or something else...
I believe your results are also true, just need to dig into them. I would perform another round of tests with DAAL and update this thread ASAP. thanks!

**Comment by rcurtin** [comment]
It's no problem, I'm happy to close as-is also if you prefer.  That you are getting different results with scikit's defaults confuses me also---if you are using `DecisionTreeClassifier` with its defaults it will build a decision tree that should be exactly equivalent to mlpack's with a minimum_leaf_size of 1 (minus small implementational differences perhaps).  That will significantly overfit the data in the vast majority of cases and you can see in the ROC curves I produced for the data that you gave that this is indeed the case.

So, I guess from my end, perhaps I am not able to reproduce what you are seeing.  It may be worth taking a look at the code I wrote and seeing if somehow it gives you different ROC curves or something.

**Comment by rcurtin** [comment]
It's been two months now---I'm going to close this for inactivity now to clean up the bug tracker.  However we can definitely reopen it if you like, just let me know.  Thanks again for your time with this one.
